### ISSUE TITLE
Make sure trajectories can "time-out" while they are being executed

### ISSUE THREAD

**Issue by ted-miller** [Yaskawa-Global/motoros2#39]
we might want to make sure trajectories can "time-out" while they are being executed.

I haven't checked the current implementation, but checking for time-tolerance should preferably be independent from trajectory execution (as in: we shouldn't just check for goal time violation at the end of the trajectory, but always).

Complexity here would be to figure out what to do when the robot is still in motion.

We'd need to initiate some sort of controlled stop, or generate a stop trajectory and replace the currently executing trajectory.

**Comment by gavanderhoorn** [comment]
This could potentially also help with situations such as reported in #41.

**Comment by jimmy-mcelwain** [comment]
A note from the documentation for [joint_trajectory_controller](http://wiki.ros.org/joint_trajectory_controller):

`Important note Even when a goal has been aborted, the controller will still attempt to execute the trajectory as best as possible.`

Also, unlike tolerances for joints during execution, `control_msgs/FollowJointTrajectory` actions do not have "live" checks for each joint trajectory. There is only a final time check at the end, after it has reached the goal state. We could possibly use the [`goal_time_tolerance`](https://github.com/ros-controls/control_msgs/blob/e787119747d7a2271e93b25133e93b392fc389af/control_msgs/action/FollowJointTrajectory.action#L30) variable, but I do not think that is it's intended purpose.

For a joint trajectory with 3 points `A`, `B`, `C`, and `D` intended to be executed at `T=0`, `T=1`, `T=2`, `T=3` respectively with `goal_time_tolerance=0.5` if beginning to execute point `C` at `T=2.6`, should we:
1. Abort trajectory and hold/try to end movement before trajectory is complete
2. Abort trajectory, but continue moving until the trajectory is complete
3. Do not abort trajectory, but abort and hold/try to end movement if point `D` is not reached before `T=3.5`
4. Do not abort trajectory, but abort if point `D` is not reached before `T=3.5`, but continue moving until the trajectory is complete

So basically:
- Are we going to continue moving until the trajectory is complete?
- Should we check at each point, whether we are within `goal_time_tolerance` of that point, or should we check if we are within `goal_time_tolerance` of only the final point?
- Also, when should we make these checks? At the beginning of processing each user-supplied point? While processing each calculated increment?

My preference would actually be to close this issue, since it seems like convention is to keep it running until the trajectory is complete, even if it is aborted. But if we are talking about aborting/stopping during the trajectory if it clearly gets "stuck" (e.g. it hasn't moved in the middle of the trajectory for 10 seconds) as some sort of error prevention measure, then I get it though.

**Comment by gavanderhoorn** [comment]
IIRC, when we discussed this the idea was to use `time_from_start` of the last `JointTrajectoryPoint` as encoding the total duration of the trajectory.

Execution of a goal longer than that should result in an `ABORT`, as clearly something is wrong -- except perhaps when an FSU is active and scaling velocities.

As to what should happen, in my opinion: `ABORT`, always (of the motion *and* the action goal).

The Yaskawa controller already tracks and monitors execution of motions. Tracking errors (on the servo level) would cause motion to be aborted. Safety errors (FSU or even just an e-stop) would cause motion to be aborted.

MotoROS2 currently aborts active goals if motion is stopped by the Yaskawa motion controller.

I'm not completely sure anymore whether when we logged this ticket we had that implemented, but regardless: anything which causes trajectory execution to be longer than `points[n-1].time_from_start +- goal_time_tolerance` would seem like it should be treated like an error and should result in a full stop + an `ABORT` of the action goal.

@jimmy-mcelwain wrote:
> A note from the documentation for [joint_trajectory_controller](http://wiki.ros.org/joint_trajectory_controller):
> 
> `Important note Even when a goal has been aborted, the controller will still attempt to execute the trajectory as best as possible.`

IIRC, that note was added as a result of the discussion in ros-controls/ros_controllers#48. Cancelling a goal does result in a stop, see [the implementation](https://github.com/ros-controls/ros_controllers/blob/c2348e85abf35cf33cf654a84d61db206abe9a73/joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h#L638-L660). The note isn't very clear about it, but it describes the behaviour of the controller when it detects a violation of any tolerances/constraints. And I can't be sure, but I also believe it's outdated, as [tolerance violations do abort goals according to this](https://github.com/ros-controls/ros_controllers/blob/c2348e85abf35cf33cf654a84d61db206abe9a73/joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h#L390-L414).

---

As I wrote in https://github.com/Yaskawa-Global/motoros2/issues/39#issuecomment-1573143291 this could potentially also help with #41, but ideally we shouldn't need to do anything special in case an action client 'disappears'. Unfortunately there doesn't appear to be an easy way to check for 'liveliness' of an action client, but after a goal has completed the server should return to idle. Why that didn't happen for @ted-miller in #41 I can't say.

**Comment by gavanderhoorn** [comment]
@gavanderhoorn wrote:
> And I can't be sure, but I also believe it's outdated, as [tolerance violations do abort goals according to this](https://github.com/ros-controls/ros_controllers/blob/c2348e85abf35cf33cf654a84d61db206abe9a73/joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h#L390-L414).

or maybe not, according to the many issues still open about this (such as ros-controls/ros_controllers#461).

That doesn't match my recollection though, but apparently it's (still) a thing.

Obviously we want to *not* do that in MotoROS2.

**Comment by jimmy-mcelwain** [comment]
Thanks for the sources @gavanderhoorn. I found another source that I think supports what you said. 

>If tolerances are violated during trajectory execution, the action goal is aborted, the client is notified, and the current position is held. 

[source](https://control.ros.org/rolling/doc/ros2_controllers/joint_trajectory_controller/doc/userdoc.html)

What I linked earlier probably is outdated. I think that holding and aborting is what makes the most sense.

**Comment by jimmy-mcelwain** [comment]
> Execution of a goal longer than that should result in an ABORT, as clearly something is wrong -- except perhaps when an FSU is active and scaling velocities.

The the "except perhaps when an FSU is active" seems important.

If the FSU is active, do we still want to abort the motion eventually, but just give it extra time? 

And more importantly, right now we are determining if the FSU is active just by observing if the trajectories are being modified/the stated trajectory is not being successfully followed. If the trajectory is "off", we seem to assume that the FSU is active. I see in https://github.com/pyumr/motoros2-internal/issues/409 that we seemingly don't have a definitive way to tell if the FSU is modifying the trajectory (e.g. we can't find a cio bit for that). I could try to look into this a little bit more, but the fact that we are trying to allow FSU movements without actually definitively knowing when the FSU is modifying trajectories makes it tough.

**Comment by jimmy-mcelwain** [comment]
After diving into for a while and looking at documentation, it seems like it is true that we can't tell if the FSU is actively modifying trajectories. We can tell if the FSU has rules that are "valid" and if it *would* modify trajectories that violate some limit if that limit was exceeded, but it doesn't indicate when the limit is being exceeded/when the trajectory is being modified. 

We could 
1. Check if the FSU has any valid rules in general, and if it does, allow trajectories to run for longer. I do not like this option, especially because FSU rules can change from valid to invalid at any time. 
2. Have a "strict goal validation" boolean in the config file that -- if set to true -- would abort a trajectory that does not finish within `goal_time_tolerance` of the final point, but if set to false lets it run for longer. I like this, at least as a short term solution for this specific issue.
3. Request a way to tell if the speed is being modified by the FSU from YEC.
